### PR TITLE
[SPARK-21726][SQL] Check for structural integrity of the plan in Optimzer in test mode.

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

We have many optimization rules now in `Optimzer`. Right now we don't have any checks in the optimizer to check for the structural integrity of the plan (e.g. resolved). When debugging, it is difficult to identify which rules return invalid plans.

It would be great if in test mode, we can check whether a plan is still resolved after the execution of each rule, so we can catch rules that return invalid plans.

## How was this patch tested?

Added tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -37,6 +37,12 @@ import org.apache.spark.sql.types._
 abstract class Optimizer(sessionCatalog: SessionCatalog)
   extends RuleExecutor[LogicalPlan] {
 
+  // Check for structural integrity of the plan in test mode. Currently we only check if a plan is
+  // still resolved after the execution of each rule.
+  override protected def planChecker: Option[LogicalPlan => Boolean] = Some(
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: can we move the checking of whether this is a test in here, then this method simply returns boolean.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Thanks. I will update it.

**Comment by viirya** [comment]
Interesting, existing `PullupCorrelatedPredicates` produces unresolved plan. I'll figure out the reason.

**Comment by viirya** [comment]
The reason `PullupCorrelatedPredicates` produces unresolved plan:

The query causing the problem in `subquery_in_having.q` looks like:

    select b.key, min(b.value)
    from src b
    group by b.key
    having b.key in ( select a.key
                    from src a
                    where a.value > 'val_9' and a.value = min(b.value)
                    )
    order by b.key
    ;

The optimized plan looks like:

    'Sort [key#201 ASC NULLS FIRST], true
    +- 'Project [key#201, min(value)#204]
       +- 'Filter key#201 IN (list#200 [(value#207 = min(value)#204)])
          :  +- Project [key#206, value#207]
          :     +- Filter (value#207 > val_9)
          :        +- InMemoryRelation [key#206, value#207], true, 5, StorageLevel(disk, memory, deserialized, 1 replicas), src
          :              +- HiveTableScan [key#0, value#1], HiveTableRelation `default`.`src`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [key#0, value#1]
       +- Aggregate [key#201], [key#201, min(value#202) AS min(value)#204, min(value#202) AS min(value#202)#209]
       +- InMemoryRelation [key#201, value#202], true, 5, StorageLevel(disk, memory, deserialized, 1 replicas), src
       +- HiveTableScan [key#0, value#1], HiveTableRelation `default`.`src`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [key#0, value#1]

Before `PullupCorrelatedPredicates` rule, the subquery in `ListQuery` looks like:

    Project [key#206]
    +- Filter ((value#207 > val_9) && (value#207 = outer(min(value)#204)))
       +- InMemoryRelation [key#206, value#207], true, 5, StorageLevel(disk, memory, deserialized, 1 replicas), src
             +- HiveTableScan [key#0, value#1], HiveTableRelation `default`.`src`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [key#0, value#1]

Currently the `In` predicate in `Filter` is resolved.

After the rule, the subquery looks like:

    Project [key#206, value#207]
    +- Filter (value#207 > val_9)
       +- InMemoryRelation [key#206, value#207], true, 5, StorageLevel(disk, memory, deserialized, 1 replicas), src
             +- HiveTableScan [key#0, value#1], HiveTableRelation `default`.`src`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [key#0, value#1]

Notice `key#206` has been added into `Project` and the condition `value#207 = outer(min(value)#204)` has been pulled out to top `Filter`:

     'Filter key#201 IN (list#200 [(value#207 = min(value)#204)])

Because `In.checkInputDataTypes` checks if the size of left columns (key#201) matches the size of subquery output (key#206, value#207), so it fails and return `false` for `In.resolved`.

The unresolved `Filter` doesn't cause problem before because it will be converted to `Join` by `RewritePredicateSubquery` rule later. But it has been detected by this structural integrity check.

By modifying `In.checkInputDataTypes`, we can solve this issue. I'd submit another PR for it.

**Comment by viirya** [comment]
The PR going to fix the issue in `PullupCorrelatedPredicates` is submitted at #18968.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
#18968 is merged. This should pass the tests now.

**Comment by viirya** [comment]
Seems there are other issues caused by `RewritePredicateSubquery` rule. I'll investigate and fix it.

**Comment by viirya** [comment]
`RewritePredicateSubquery` fails structural integrity check because it can produce `Join` with conflicting attributes in its left and right plans.

I submitted #19050 to fix it.

**Comment by viirya** [comment]
#19050 is merged now. Let's see if there is any rule can fail this structural integrity check.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala [hunk 2]
```diff
@@ -93,6 +101,13 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
                 """.stripMargin)
             }
 
+            // Run the structural integrity checker against the plan after each rule.
+            if (!planChecker(result)) {
+              val message = s"After applying rule ${rule.ruleName} in batch ${batch.name}, " +
+                "the structural integrity of the plan is broken."
+              throw new TreeNodeException(result, message, null)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: move the exception throwing logics into the `planChecker `?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizerSICheckerSuite.scala [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.spark.sql.catalyst.dsl.plans._
+import org.apache.spark.sql.catalyst.errors.TreeNodeException
+import org.apache.spark.sql.catalyst.expressions.{Alias, Literal}
+import org.apache.spark.sql.catalyst.plans.PlanTest
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, OneRowRelation, Project}
+import org.apache.spark.sql.catalyst.rules._
+import org.apache.spark.sql.internal.SQLConf
+
+
+class OptimizerSICheckerkSuite extends PlanTest {
```
Comment (gatorsmile) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizerSICheckerSuite.scala]: -> OptimizerStructuralIntegrityCheckerkSuite

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala (hunk 2 above)
Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: nvm. The message also has rule and batch names.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala [hunk 4]
```diff
@@ -64,6 +64,14 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
   protected def batches: Seq[Batch]
 
   /**
+   * Defines a check function which checks for structural integrity of the plan after the execution
+   * of each rule. For example, we can check whether a plan is still resolved after each rule in
+   * `Optimizer`, so we can catch rules that return invalid plans. The check function will returns
+   * `false` if the given plan doesn't pass the structural integrity check.
+   */
+  protected def planChecker(plan: TreeType): Boolean = true
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: `planChecker ` -> `isPlanIntegral`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizerSICheckerSuite.scala (hunk 3 above)
Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizerSICheckerSuite.scala]: Ok.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala (hunk 4 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: Looks good.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala [hunk 5]
```diff
@@ -64,6 +64,14 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
   protected def batches: Seq[Batch]
 
   /**
+   * Defines a check function which checks for structural integrity of the plan after the execution
+   * of each rule. For example, we can check whether a plan is still resolved after each rule in
+   * `Optimizer`, so we can catch rules that return invalid plans. The check function will returns
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: `will returns` -> `returns`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala [hunk 6]
```diff
@@ -64,6 +64,14 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
   protected def batches: Seq[Batch]
 
   /**
+   * Defines a check function which checks for structural integrity of the plan after the execution
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: `which` -> `that`

**Comment by gatorsmile** [comment]
LGTM except two minor comments

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Comment by viirya** [comment]
Thanks @rxin @gatorsmile

**Pull request closed without merging**